### ISSUE TITLE
Use zstd instead of gzip

### ISSUE THREAD

**Issue by imbsky** [actions/cache#184]

**Comment by imbsky** [comment]
We need to wait for this issue to be closed. https://github.com/actions/virtual-environments/issues/89

**Comment by imbsky** [comment]
#6 has some useful information from @krcroft and @hugovk.

These benchmarks are really useful to consider this: https://github.com/actions/cache/issues/6#issuecomment-551012192 https://github.com/actions/cache/issues/6#issuecomment-551101574

**Comment by imbsky** [comment]
https://github.com/actions/virtual-environments/pull/414

**Comment by imbsky** [comment]
PR has been merged upstream.

**Comment by imbsky** [comment]
@joshmgross Now this is not blocked by anything. Do you have the bandwidth to do this?

**Comment by joshmgross** [comment]
Thanks @imbsky for following up and getting https://github.com/actions/virtual-environments/pull/414 merged.

We'll need to wait for that to get added to the `macOS` image and then for the latest image updates to roll out with `zstd` installed. Typically that's a monthly update, but I'll follow up with the team to verify.

Once https://github.com/actions/virtual-environments/issues/89 is closed I'll take a look at implementing this.
Hopefully should be a small fix, but need to handle back-compat and any self-hosted runners that may not have `zstd` installed.

**Comment by imbsky** [comment]
Yes, indeed, it's a little troublesome to handle self-hosted runners.

**Comment by imbsky** [comment]
I wrote code that ignores backward compatibility. I can do the work for you, or you can refer to this and add changes such as backwards compatibility.
https://github.com/imbsky/cache/commit/e8341fc6fe7b2fa38c1b517c2bc57c5fd987b831

**Comment by imbsky** [comment]
zstd has also been added to macOS. It seems to be deployed in a week or two.

**Comment by imbsky** [comment]
Hmm... My code doesn't seem to work in some environments. When using zstd from tar, the behavior is unstable, so you may want to break it up into separate operations.

**Comment by joshmgross** [comment]
@imbsky Some environments will use BSD tar and some will use GNU tar, so make sure the arguments you specify ones that are recognized by both.

For archive creation, you'd want to replace the `-xz` with `-x` and add the `--zstd` option.
I think for archive extraction you can just remove the `z` from `-cz`, as both GNU and BSD tar say they will automatically decompress based on the file.

See https://www.freebsd.org/cgi/man.cgi?tar(1) and https://www.gnu.org/software/tar/manual/tar.html#SEC135 for more info.

**Comment by imbsky** [comment]
BSD tar in macOS and Windows seems to be difficult to use with zstd. I tried several ways but all failed. It's probably a good idea to install Gnu tar or separate the tar and zstd operations. The latter is probably smarter.

**Comment by krcroft** [comment]
To maximize cache performance on macOS, I've used GNU tar plus a `tar`-wrapper script to run zstd under the hood (GitHub's cache module calls this script because it shows up first in PATH, and has been working reliably for the last couple months):

**Compression:** `gtar --use-compress-program="zstd -T0 --adapt --long=31 --no-progress -v" <existing arguments>`, minus the `-z`

**Decompression:** `gtar --use-compress-program="zstd --long=31 --no-progress -v -d" <existing arguments>`, minus the `-z`

Flag descriptions:
- `T0`, allocate all cores/threads to compression; might as well make the most of the available cores because there's nothing else that can use them during this time anyway (except background/cron jobs).  This will give you close to 4x performance on our GitHub VMs.
- `--adapt` bounds the compression level to the edge of what the underlying IO subsystem can sustain (both the read and write rates).  
   This prioritizes completion time over compression rates.  In practice on solid state or pure-memory systems, this means moving 1+ GBs at LZ4 type compression rates (and lighting up a handful of cores); but on GitHub's slow-ish VMs (moving perhaps <100MB/s in either direction), zstd can ratchet up compression rates quite high (and across all four cores), essentially for free while the IO subsystem runs at peak rates.
- `--long=31` will find and compress redundancy in runs of data of up to 2GB (far more than gzip's 32KB window size!).  The VMs have the memory, and we're already maximizing speeds - so we might as well wring out as much redundancy in the cached data as we can.
- `--no-progress` avoids printing too much on-the-fly status details, which can otherwise fill pages of logs during large compression runs. 
- `-v` still prints the overall compression and rates achieved at the end, which is useful.

Give it a whirl.. and watch your cache fly! 🏃

**Comment by imbsky** [comment]
"--adapt" and "--no-progress" aren't supported in versions that can be installed on Ubuntu and Windows, so I don't think We will be able to use them for a while.

**Comment by imbsky** [comment]
Running tar command from Node.js will fail. But it doesn't seem to fail if you don't run it from Node.js. I'm not an expert on Node.js so I don't know why, but do you know this behavior?

**Comment by krcroft** [comment]
It looks like the entire compress program argument string is being treated as the executable (`zstd -T0 --long=31: not found`), instead of just the first word. 

try this for the creation step in your macos workflow (homebrew's zstd version is 1.4.4, which supports adaptive compression):

run: tar -c node_modules  | zstd -T0 --adapt --long=31 -f -o node_modules.tar.zst

**Comment by imbsky** [comment]
This only happens when running from Node.js and works fine if called directly, as `run: tar --use-compress-program="zstd -T0 --long=31" -cf node_modules.tar.zst node_modules`.

**Comment by imbsky** [comment]
Okay, we need to wait for this to be merged. https://github.com/actions/toolkit/pull/285

**Comment by imbsky** [comment]
@joshmgross Hmm... It doesn't seem to matter. What is the difference between `exec `and `child_process.spawn`?

**Comment by imbsky** [comment]
Oh, okay. I finally succeeded in executing that on all platforms.

**Comment by imbsky** [comment]
For some reason macOS and Windows are very slow, but I don't know if this is a problem with the Actions server or a code problem.

**Comment by joshmgross** [comment]
>What is the difference between exec and child_process.spawn?

`exec` internally calls `child_process.spawn`, but has a lot of setup to make it more user-friendly and handle cross-platform issues (i.e. handling Windows issues, outputs, etc). 
See https://github.com/actions/toolkit/blob/master/packages/exec/src/toolrunner.ts#L436

**Comment by imbsky** [comment]
It looks pretty good!

**Comment by imbsky** [comment]
@joshmgross By the way, I think that just bumping the major version is enough for backwards compatibility. What do you think?

**Comment by joshmgross** [comment]
>By the way, I think that just bumping the major version is enough for backward compatibility

Yeah, that's definitely an option, but the main concern is that we would fragment the usage of this action as not everyone will immediately switch to `v2`. If it is a breaking change, then that would be the correct move. Ideally, we could combine that breaking change with other ones (i.e. multi-path support) to incentivize moving to the latest version and avoiding have extra major versions.

**Comment by imbsky** [comment]
@joshmgross That idea sounds very good to me. Are you currently working on multi-path support?

**Comment by joshmgross** [comment]
>Are you currently working on multi-path support?

Not at this time, no. But it will likely be the next big feature we add along with a `@actions/cache` npm package for actions authors

**Comment by imbsky** [comment]
@joshmgross Great. Please let me know if there is anything I can do for you. I don't know what I can do, but I don't like to only complain and request, so I will do what I can if you need me 🙂

**Comment by imbsky** [comment]
In my OCaml project, the compression time has been reduced from 6 minutes to 2 minutes! Amazing speed!

**Comment by joshmgross** [comment]
Thanks @imbsky, much appreciated.

Right now that specific work is blocked by https://github.com/actions/toolkit/issues/184, and it will require some backend changes to allow action inputs to be arrays rather than just strings.

**Comment by imbsky** [comment]
It works fine in my environment, but for some reason, it takes a long time in Actions macOS and Windows environments. (It doesn't fail with an error.)

**Comment by imbsky** [comment]
https://github.com/imbsky/actions-cache-playground/actions/runs/47209108

**Comment by imbsky** [comment]
@joshmgross I'm sorry. I forgot to reply to your reply. Okay, I'll wait until it's changed 🙂

**Comment by imbsky** [comment]
https://github.com/imbsky/actions-cache-playground/runs/488906222

**Comment by maxim-lobanov** [comment]
@imbsky , Issue with default pre-installed packages on macOS (2.2 GB on Yarn by default and some default NPM packages) is tracked by https://github.com/actions/cache/issues/184. It is on our board and will be fixed soon. No ETA for now.

**Comment by imbsky** [comment]
Great, thanks.

**Comment by annulen** [comment]
I have large cache file and I'd like to use xz (i.e., replace `tar -czf` when cache is archived to `tar -cJf`). Should I create another issue, or this one can be generalized to supporting multiple compression algorithms?

**Comment by imbsky** [comment]
@annulen Maybe you want to use bzip2 instead of gzip?

**Comment by annulen** [comment]
No, compression of bzip2 is inferior to xz (which uses LZMA algorithm). Nowadays bzip2 is usually considered to be obsolete algorithm, it's only strong point is lower memory usage on compression, which isn't relevant at all when you have 7 Gb of RAM.

**Comment by imbsky** [comment]
Yeah, I just wanted to know what you thought. For example, what kind of compression algorithm do you want to use?

**Comment by annulen** [comment]
As I said, I want to use `xz` (`-J` option of `tar`)

**Comment by imbsky** [comment]
Oh, sorry, I misunderstood the option. It's easy to do.

**Comment by imbsky** [comment]
But if you cache a file that is large that you need to use xz instead of zstd, you will probably experience cache thrashing.

**Comment by annulen** [comment]
That archive is about 2Gb, so I should still have 3Gb available for other jobs.

**Comment by krcroft** [comment]
@annulen , @imbsky -- is GitHub's cache going to configurable by the user? (ie: zstd / gzip / xz / .. ) along with the compression level?

I remember @chrispat preferring a simpler one-size-fits-all solution, and I think zstd fits this bill. zstd at -19 (or --ultra -22) plus 2GB window-size should roughly match or exceed the best 'xz' can do (compression-wise), plus be several-fold faster time-wise.

If we are leaning toward user-configurability, perhaps one user-configurable compression option would suffice:  `**balanced**` or ``.  `**minimum-size**`.

- **balanced**  would be the unstated-default, and use the current cache as-is (default zstd level of compression and in the future use `-adapt`).
- **minimum-size** would crank up the compression to `-19` (all other settings as-is). 

You might wonder if a **minimum-time** (or **fastest**) option is needed - but I don't think so.  The defaults right now exceed what the underlying IO system can provide; so lowering the compression level further will only bloat the result and add IO write time.  Likewise, `-adapt` in the future will always go as fast as the IO subsystems, and that's a hard bottleneck.

**Comment by imbsky** [comment]
Yes, so I was just trying to post a comment that we should open a new issue to consider it.

**Comment by annulen** [comment]
I also see that there is a request for no compression option in #161

**Comment by krcroft** [comment]
zstd is very unique among compressors in that it covers a very wide range: multi-GB/s (snappy/lzo/lz4) up in the mid-range (zip/bzip2/rar/gzip), to the high-range (xz/lizard/lzma), and it matches or exceeds them performance wise. 

Prior to zstd, a trio of best-in-class compressors: say **lz4** + **zopfli/brotli/gzip** + **xz** might have fit the bill. But zstd runs past all of them in their weight-classes, so we don't have to take on this complexity in the back-end (or burden users with it..).

**Comment by annulen** [comment]
>But if you cache a file that is large that you need to use xz instead of zstd, you will probably experience cache thrashing.

Now I get what you mean: large files with different cache keys, produced by repeated runs of same job, will wash away caches of other jobs instead of superceding each other.

**Comment by imbsky** [comment]
This is no longer blocked.

**Comment by imbsky** [comment]
There has been some progress on this recently.
